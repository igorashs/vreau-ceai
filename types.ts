import FormidableLib from 'formidable';
import { ValidationError } from 'joi';

export type Product = {
  _id: string;
  name: string;
  category_id: string;
  src: string;
  price: number;
  quantity: number;
  total_quantity: number;
  description: string;
  recommend: boolean;
};

export type ProductWithCategory = Product & {
  category_id: { name: string };
};

export type ErrorDetail = {
  message: string;
  name?: string;
};

export type ProductName = {
  name: string;
};

export type ProductNameErrorDetail = ErrorDetail & {
  name?: keyof ProductName;
};

export type ProductFields = ProductName & {
  src?: string;
  price: number;
  quantity: number;
  total_quantity: number;
  description: string;
  recommend: boolean;
  category_id: string;
};

export type Category = {
  _id: string;
  name: string;
};

export type ProductErrorDetail = ErrorDetail & {
  name?: keyof ProductFields;
};

export type ApiResponse = {
  success: boolean;
  message: string;
  errors?: ErrorDetail[];
};

export type User = {
  _id: string;
  name: string;
  email: string;
  isManager: boolean;
};

export type UserEmail = {
  email: string;
};

export type UserEmailErrorDetail = ErrorDetail & {
  name?: keyof UserEmail;
};

export type UserLogin = UserEmail & { password: string };

export type UserSignup = UserEmail & {
  name: string;
  password: string;
  repeat_password: string;
};

export type UserPermissions = { isManager: boolean };

export type LabelMessage = { success: boolean; message: string };

export type UserAuth = {
  _id: string;
  name: string;
  isAdmin: boolean;
  isManager: boolean;
};

export type UserSession = {
  isAuth: boolean;
  user: UserAuth | null;
  needRefresh: boolean;
};

type dataType = 'data' | 'error' | 'field' | 'fileBegin' | 'file' | 'progress';

export declare class Formidable extends FormidableLib {
  emit(name: dataType, data: FormidableLib.EmitData | ValidationError): void;
}

export type CategoryName = {
  name: string;
};

export type CategoryNameErrorDetail = ErrorDetail & {
  name?: keyof CategoryName;
};

export type OrderSubmit = {
  tel: string;
  address: string;
};

export type OrderItem = {
  product_id: string;
  count: number;
};

export type OrderStatus = {
  status: 'processing' | 'inDelivery' | 'canceled' | 'completed';
};

export type OrderStatusErrorDetail = ErrorDetail & {
  name?: keyof OrderStatus;
};

export type OrderNumber = {
  number: string;
};

export type OrderNumberErrorDetail = ErrorDetail & {
  name?: keyof OrderNumber;
};

export type OrderFields = {
  info: OrderSubmit;
  items: OrderItem[];
};

export type Order = {
  _id: string;
  user: {
    name: string;
    email: string;
  };
  number: string;
  total_price: number;
  status: 'processing' | 'inDelivery' | 'canceled' | 'completed';
  items: {
    count: number;
    product: { name: string; price: number; quantity: number };
  }[];
  address: string;
  tel: string;
  orderedAt: Date;
  completedAt: Date;
};

export type UserOrders = { orders: Order[]; count: number };

export type ManagementOrders = { orders: Order[]; count: number };
