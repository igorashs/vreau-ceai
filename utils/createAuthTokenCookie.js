import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
