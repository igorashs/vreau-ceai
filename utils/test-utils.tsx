import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
