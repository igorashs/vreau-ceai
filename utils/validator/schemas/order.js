import Joi from 'joi';

export const orderSubmitSchema = Joi.object({
  tel: Joi.string()
    .pattern(/^0\d{8}$/)
    .required()
    .messages({
      'string.empty': 'telefonul este obligatoriu',
      'any.required': 'telefonul este obligatoriu',
      'string.pattern.base': 'telefonul este invalid'
    }),
  address: Joi.string().trim().max(200).messages({
    'string.empty': 'adresa este obligatorie',
    'any.required': 'adresa este obligatorie',
    'string.max': 'adresa este prea lungÄƒ'
  })
});

export const orderItemsSchema = Joi.array().items(
  Joi.object({
    product_id: Joi.string().required(),
    count: Joi.number().min(1).required()
  })
);
