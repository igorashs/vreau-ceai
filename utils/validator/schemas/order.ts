import Joi from 'joi';
import { OrderItem, OrderNumber, OrderStatus, OrderSubmit } from 'types';

export const orderMessages = {
  tel: {
    invalid: 'telefonul este invalid',
    required: 'telefonul este obligatoriu',
  },

  address: {
    max: 'adresa este prea lungă',
    required: 'adresa este obligatorie',
  },

  number: {
    length: 'numărul trebuie să aibă exact 16 caractere',
    required: 'numărul este obligatoriu',
    wrong: 'numărul este greșit',
  },

  status: {
    invalid: 'status greșit',
    required: 'statusul este obligatoriu',
  },
};

export const orderSubmitSchema = Joi.object<OrderSubmit>({
  tel: Joi.string()
    .pattern(/^0\d{8}$/)
    .required()
    .messages({
      'string.empty': orderMessages.tel.required,
      'any.required': orderMessages.tel.required,
      'string.pattern.base': orderMessages.tel.invalid,
    }),
  address: Joi.string().trim().max(200).messages({
    'string.empty': orderMessages.address.required,
    'any.required': orderMessages.address.required,
    'string.max': orderMessages.address.max,
  }),
});

export const orderItemsSchema = Joi.array().items(
  Joi.object<OrderItem>({
    product_id: Joi.string().required(),
    count: Joi.number().min(1).required(),
  }),
);

export const orderNumberSchema = Joi.object<OrderNumber>({
  number: Joi.string().length(16).required().messages({
    'string.length': orderMessages.number.length,
    'string.empty': orderMessages.number.required,
    'any.required': orderMessages.number.required,
  }),
});

export const orderStatusSchema = Joi.object<OrderStatus>({
  status: Joi.string()
    .valid('processing', 'inDelivery', 'canceled', 'completed')
    .required()
    .messages({
      'string.empty': orderMessages.status.required,
      'any.required': orderMessages.status.required,
      'any.only': orderMessages.status.invalid,
    }),
});
