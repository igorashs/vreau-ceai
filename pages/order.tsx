import withBaseLayout from '@/layouts/BaseLayout';
import Head from 'next/head';
import styled from 'styled-components';
import Form, { FormAction } from '@/shared/Form';
import { withSessionServerSideProps } from '@/utils/withSession';
import { useForm } from 'react-hook-form';
import { joiResolver } from '@hookform/resolvers/joi';
import { orderSubmitSchema } from '@/utils/validator/schemas/order';
import TextField from '@/shared/TextField';
import Button from '@/shared/Button';
import { useCart } from 'contexts/CartContext';
import { createOrder } from 'services/ceaiApi';
import { useState } from 'react';
import Label from '@/shared/Label';
import { LabelMessage, OrderSubmit } from 'types';

const FormWrapper = styled.div`
  max-width: var(--max-input-width);
  width: var(--max-input-width);
`;

const Wrapper = styled.div`
  display: grid;
  gap: var(--baseline);
`;

export default function Order() {
  const {
    register,
    handleSubmit,
    setError,
    formState: { errors },
  } = useForm({
    mode: 'onChange',
    resolver: joiResolver(orderSubmitSchema),
  });

  const [cart, cartDispatch] = useCart();
  const [label, setLabel] = useState<LabelMessage>();

  const onSubmit = async (data: OrderSubmit) => {
    const items = cart.items.map(({ product, count }) => ({
      product_id: product._id,
      count,
    }));

    const res = await createOrder({
      info: data,
      items,
    });

    if (res.success) {
      setLabel({
        success: true,
        message: `comanda cu nr. ${res.number} a fost trimisă, un operator vă va contacta încurând.`,
      });

      cartDispatch({ type: 'clear-cart' });
    }

    if (res?.errors) {
      res.errors.forEach((error) => {
        const { message, name } = error;
        if (name) setError(name, { message });
      });
    } else if (!res.success) {
      setLabel({
        success: false,
        message: 'comanda nu a fost trimisă, goliți coșul și încercați din nou',
      });
    }
  };

  return (
    <>
      <Head>
        <title>Comandă</title>
      </Head>

      <Wrapper>
        {cart?.itemsCount <= 0 ? (
          !label && <p>coșul este gol :(</p>
        ) : (
          <>
            <h4>Comandă</h4>

            <FormWrapper>
              <Form onSubmit={handleSubmit(onSubmit)}>
                <TextField
                  name="tel"
                  label="telefon"
                  placeholder="ex: 061234567"
                  error={errors?.tel?.message}
                  passRef={register}
                  type="tel"
                />
                <TextField
                  name="address"
                  label="adresa"
                  placeholder="ex: or. Strășeni str. Mihai Em. nr. 28"
                  error={errors?.address?.message}
                  passRef={register}
                  type="text"
                />
                <FormAction justify="flex-end">
                  <Button disabled={cart.itemsCount <= 0}>trimite</Button>
                </FormAction>
              </Form>
            </FormWrapper>
          </>
        )}

        {label && (
          <Label error={!label.success} success={label.success}>
            {label.message}
          </Label>
        )}
      </Wrapper>
    </>
  );
}

export const getServerSideProps = withSessionServerSideProps(
  async ({ req }) => {
    const { isAuth } = req.session;

    if (!isAuth) {
      return {
        redirect: {
          destination: '/',
          permanent: false,
        },
      };
    }

    return { props: {} };
  },
);

Order.withLayout = withBaseLayout;
