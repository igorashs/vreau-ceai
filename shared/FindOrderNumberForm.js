import { TextField } from '@/shared/TextField';
import { Button } from '@/shared/Button';
import { useForm } from 'react-hook-form';
import { joiResolver } from '@hookform/resolvers/dist/ie11/joi';
import { orderNumberSchema } from '@/utils/validator/schemas/order';
import { Form, FormAction } from '@/shared/Form';

export function FindOrderNumberForm({ onFindOrderSubmit }) {
  const {
    register,
    handleSubmit,
    setError,
    formState: { errors }
  } = useForm({
    mode: 'onChange',
    resolver: joiResolver(orderNumberSchema)
  });

  const onSubmit = async (data) => {
    const errors = await onFindOrderSubmit(data);

    if (errors) {
      errors.forEach((error) => {
        const { message, name } = error;
        setError(name, { message });
      });
    }
  };

  return (
    <Form onSubmit={handleSubmit(onSubmit)}>
      <TextField
        name="number"
        label="numărul comenzii"
        error={errors?.number?.message}
        passRef={register}
        type="text"
      />
      <FormAction justify="flex-end">
        <Button>caută</Button>
      </FormAction>
    </Form>
  );
}
