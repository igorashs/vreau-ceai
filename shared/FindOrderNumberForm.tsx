import { TextField } from '@/shared/TextField';
import Button from '@/shared/Button';
import { useForm } from 'react-hook-form';
import { joiResolver } from '@hookform/resolvers/dist/ie11/joi';
import { orderNumberSchema } from '@/utils/validator/schemas/order';
import { Form, FormAction } from '@/shared/Form';
import { OrderNumber } from 'types';

type InputsErrors = { message: string; name: 'number' }[];

interface FindOrderNumberFormProps {
  onFindOrderSubmit: (data: OrderNumber) => InputsErrors;
}

export const FindOrderNumberForm = ({
  onFindOrderSubmit,
}: FindOrderNumberFormProps) => {
  const {
    register,
    handleSubmit,
    setError,
    formState: { errors },
  } = useForm({
    mode: 'onChange',
    resolver: joiResolver(orderNumberSchema),
  });

  const onSubmit = async (data: OrderNumber) => {
    const submitErrors = await onFindOrderSubmit(data);

    if (submitErrors) {
      submitErrors.forEach((error) => {
        const { message, name } = error;
        setError(name, { message });
      });
    }
  };

  return (
    <Form onSubmit={handleSubmit(onSubmit)}>
      <TextField
        name="number"
        label="numărul comenzii"
        error={errors?.number?.message}
        passRef={register}
        type="text"
      />
      <FormAction justify="flex-end">
        <Button>caută</Button>
      </FormAction>
    </Form>
  );
};
