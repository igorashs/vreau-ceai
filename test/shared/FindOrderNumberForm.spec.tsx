import { fireEvent, render, screen, waitFor } from '@/utils/test-utils';
import FindOrderNumberForm from '@/shared/FindOrderNumberForm';
import { OrderNumberErrorDetail } from 'types';
import { orderMessages } from '@/utils/validator/schemas/order';

const onFindOrderSubmitMock = jest.fn(
  (): Promise<OrderNumberErrorDetail[] | undefined> =>
    Promise.resolve(undefined),
);

describe('FindOrderNumberForm', () => {
  it('renders correctly', () => {
    const { container, getByRole } = render(
      <FindOrderNumberForm onFindOrderSubmit={onFindOrderSubmitMock} />,
    );

    expect(getByRole('textbox')).toHaveValue('');
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('number validation', () => {
    beforeEach(() =>
      render(<FindOrderNumberForm onFindOrderSubmit={onFindOrderSubmitMock} />),
    );

    it('should display required error', async () => {
      const { getByRole, getByText } = screen;

      fireEvent.submit(getByRole('button'));

      await waitFor(() =>
        expect(getByText(orderMessages.number.required)).toBeVisible(),
      );

      expect(onFindOrderSubmitMock).not.toBeCalled();
    });

    it('should display length error', async () => {
      const { getByRole, getByText } = screen;

      fireEvent.change(getByRole('textbox'), { target: { value: '123asd' } });
      fireEvent.submit(getByRole('button'));

      await waitFor(() =>
        expect(getByText(orderMessages.number.length)).toBeVisible(),
      );

      expect(onFindOrderSubmitMock).not.toBeCalled();
    });
  });

  describe('success submission', () => {
    it('should reset the form', async () => {
      const { getByRole, getByText } = render(
        <FindOrderNumberForm onFindOrderSubmit={onFindOrderSubmitMock} />,
      );

      const number = '0JTNSLRDKKIAH91F';

      fireEvent.submit(getByRole('button'));
      fireEvent.change(getByRole('textbox'), {
        target: { value: number },
      });
      fireEvent.submit(getByRole('button'));

      await waitFor(() => expect(onFindOrderSubmitMock).toBeCalledTimes(1));
      expect(onFindOrderSubmitMock).toBeCalledWith({ number });
      expect(getByRole('textbox')).toHaveValue(number);
      expect(getByText('numÄƒrul comenzii')).toBeVisible();
    });
  });

  describe('server validation', () => {
    it('should display length error', async () => {
      // pretend user bypass client validation
      onFindOrderSubmitMock.mockResolvedValueOnce([
        { message: orderMessages.number.length, name: 'number' },
      ]);

      const { getByRole, getByText } = render(
        <FindOrderNumberForm onFindOrderSubmit={onFindOrderSubmitMock} />,
      );

      fireEvent.change(getByRole('textbox'), {
        target: { value: '0JTNSLRDKKIAH91F' },
      });
      fireEvent.submit(getByRole('button'));

      await waitFor(() => expect(onFindOrderSubmitMock).toBeCalledTimes(1));
      expect(onFindOrderSubmitMock).toBeCalledWith({
        number: '0JTNSLRDKKIAH91F',
      });
      expect(getByText(orderMessages.number.length)).toBeVisible();
    });
  });
});
