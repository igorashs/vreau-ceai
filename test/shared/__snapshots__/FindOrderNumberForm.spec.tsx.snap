// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FindOrderNumberForm renders correctly 1`] = `
.c2 {
  font-weight: 500;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 7px;
}

.c3 {
  width: 100%;
  background-color: var(--input);
  border: 1px solid var(--layout-dark);
  border-radius: 4px;
  padding: calc(var(--baseline) / 4) 0 calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  color: var(--text-dark);
}

.c3::-webkit-outer-spin-button,
.c3::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c3[type='number'] {
  -moz-appearance: textfield;
}

.c5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 7px;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c5:hover {
  background-color: var(--accent);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: calc(var(--baseline) / 2);
}

.c4 {
  margin-top: calc(var(--baseline) / 2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

<form
  class="c0"
>
  <div
    class="c1"
  >
    <label
      class="c2"
      for="number"
    >
      numărul comenzii
    </label>
    <input
      class="c3"
      id="number"
      name="number"
      type="text"
    />
  </div>
  <div
    class="c4"
  >
    <button
      class="c5"
    >
      caută
    </button>
  </div>
</form>
`;
